File: leetcode-js/106-convert-binary-tree-from-inorder-and-postorder-traversal.js
function TreeNode(val) {
  this.val = val;
  this.left = this.right = null;
}

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */
var buildTree = function(inorder, postorder) {
    if (!inorder.length) {
        return null;
    }

    return _buildTree(postorder.reverse(), 0, postorder.length, inorder.reverse(), 0, inorder.length);
};

function _buildTree(preorder, preStart, preEnd, inorder, inStart, inEnd) {
    'use strict';

    if (preStart === preEnd) {
        return null;
    }

    const val = preorder[preStart];
    const index = inorder.indexOf(val, inStart);
    const node = new TreeNode(val);

    node.right = _buildTree(preorder, preStart + 1, preStart + (index - inStart) + 1, inorder, inStart, index);
    node.left = _buildTree(preorder, preStart + (index - inStart) + 1, preEnd, inorder, index + 1, inEnd);

    return node;
}

console.log(buildTree([1,3,2], [3,2,1]));

File: leetcode-java/106-convert-binary-tree-from-inorder-and-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();

        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }

        return build(map, inorder, postorder, 0, inorder.length - 1, 0, postorder.length - 1);
    }

    TreeNode build(Map<Integer, Integer> map, int[] inorder, int[] postorder, int inStart, int inEnd, int poStart, int poEnd) {
        if (inEnd - inStart < 0) {
            return null;
        }

        TreeNode node = new TreeNode(postorder[poEnd]);
        int index = map.get(postorder[poEnd]);
        int len = inEnd - index;

        node.left = build(map, inorder, postorder, inStart, index - 1, poStart, poEnd - len - 1);
        node.right = build(map, inorder, postorder, index + 1, inEnd, poEnd - len, poEnd - 1);

        return node;
    }
}
